### PR TITLE
make fromOrdinal public and always available

### PR DESCRIPTION
**Pull request by bishabosha**
`fromOrdinal` exists on the companion object of an enum and is always defined, even if there are no value cases

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala [hunk 1]
```diff
@@ -154,22 +154,24 @@ object DesugarEnums {
...
-          CaseDef(Literal(Constant(i)), EmptyTree, enumValue)
-        ) ::: defaultCase :: Nil
-        val fromOrdinalDef = DefDef(nme.fromOrdinalDollar, Nil, List(param(nme.ordinalDollar_, defn.IntType) :: Nil),
-          rawRef(enumClass.typeRef), Match(Ident(nme.ordinalDollar_), valueCases))
-            .withFlags(Synthetic | Private)
-        fromOrdinalDef :: Nil
-
-    scaffolding ::: valueCtor ::: byOrdinal
+        def default(ordinal: Tree) =
+          CaseDef(Ident(nme.x_0), EmptyTree, throwArg(ordinal))
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala:164]: If the name isn't used then wildcard is more appropriate:
```suggestion
          CaseDef(Ident(nme.WILDCARD), EmptyTree, throwArg(ordinal))
```

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala]: thanks for the tip

**Comment by bishabosha** [comment]
tests are passing @sjrd is it correct that the `fromOrdinal` method should always be generated

**Comment by sjrd** [comment]
TBH I don't know. I have no recollection of what we discussed about `fromOrdinal`, and I haven't written down anything about that in my notes :(

**Comment by bishabosha** [comment]
Perhaps it was only if there there is at least one value :/ but I suppose this is meant to be low level for serialisation so it may be simpler to say its always there

**Comment by sjrd** [comment]
> Perhaps it was only if there there is at least one value :/

No, it's probably always. The other possibilities do not make sense.

What I'm not sure is whether we actually decided to make this public. But if we did, then it should definitely be always.

**Comment by bishabosha** [comment]
> What I'm not sure is whether we actually decided to make this public.

its supposed to be the method thats available for an arbitrary serialisation framework so I would assume external access is required

**Comment by bishabosha** [comment]
what should happen with this?

**Review by sjrd — APPROVED** [review]
I guess it should be merged.

**Pull request merged**